frontend/src/lib/DeltaParser.ts

```diff
@@ -28,14 +28,24 @@ import { MetricsManager } from "lib/MetricsManager"
 import { requireNonNull } from "lib/utils"
 
 type Container = "main" | "sidebar"
+// The actual protobuf Element.proto
 export type SimpleElement = ImmutableMap<string, any>
+// A list of nodes to render (e.g. main, sidebar, st.container, st.column)
 export interface BlockElement extends List<ReportElement> {}
 
+// Pointers to the two root nodes of the element trees.
 export interface Elements {
   main: BlockElement
   sidebar: BlockElement
 }
 
+/**
+ * A node of the element tree, representing a single thing to render: {
+ *  element: SimpleElement | BlockElement
+ *  reportId: string
+ *  metadata: IForwardMsgMetadata
+ * }
+ */
```

Comment:
Should we put ☝️ into an interface and declare the map with that instead of any? Mostly because I see us creating a report element below without any tie-in

Reply:
Yeah... I'd like this to be typed, but ImmutableMap doesn't play that well with typing afaict, which is why I just put it in a comment. Some thoughts on pros/cons of ImmutableJS here:

- Reasons to use ImmutableMap:
    - Immutability guarantee is nice
    - `getIn([0, "element", 2])` is nice for tree traverse, we'd have to implement it ourselves otherwise
- Reasons not to use ImmutableMap:
    - No type safety
    - Weird pattern: reading `object.get('property')` is harder than `object.property`
    - No autocomplete =(
    - Debugging is harder (`console.log(ImmutableMap)` is worthless)

What the internet has to say about typing immutable maps (tldr no easy solution):
*https://blog.mayflower.de/6630-typescript-redux-immutablejs.html,
*https://github.com/immutable-js/immutable-js/issues/683

Reply:
I did not realize immutable did not like types. I believe I saw that ReportElements will be go into state where immutability would be great and that we do have to traverse the tree, it makes sense to be untyped in favor of immutability